File: dailyQuestion/2020-05/05-30/python/solution_violence.py
# 暴力题解法
def largestRectangleArea(heights: [int]) -> int:
    """
    柱状图中最大的矩形
    :param heights:
    :return:
    """
    result = 0
    for i in range(len(heights)):
        height = heights[i]
        left, right = i, i
        while left - 1 >= 0 and heights[left - 1] >= height:
            left -= 1
        while right + 1 < len(heights) and heights[right + 1] >= height:
            right += 1
        result = max(result, (right - left + 1) * height)

    return result


if __name__ == "__main__":
    heights = [2, 1, 5, 6, 2, 3]
    result = largestRectangleArea(heights)
    print(result)

File: dailyQuestion/2020-05/05-30/golang/solution_violence.go
package main

import (
	"fmt"
)

func main() {
	heights := []int{6, 7, 5, 2, 4, 5, 9, 3}
	result := largestRectangleArea(heights)
	fmt.Println(result)

}

func largestRectangleArea(heights []int) int {
	result := 0
	n := len(heights)
	for i := 0; i < n; i++ {
		left, right := i, i
		for ; left-1 >= 0 && heights[left-1] >= heights[i]; left-- {

		}
		for ; right+1 < n && heights[right+1] > heights[i]; right++ {

		}
		result = max(result, (right-left+1)*heights[i])
	}
	return result
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}
